Document admin bootstrapping and recovery (#31039)
keycloak/keycloak: docs/guides/server/bootstrap-admin-recovery.adoc

Reviewer: It'd be also good to add a section for bootstrapping at server startup. Explaining the difference from the dedicated command (that on startup master realm must on exist, making the dedicated command the only suitable option for recovery).

We should also add some note that the dedicated command requires the server to be stopped and that it requires the same args as start (db options etc.).

Reviewer: > We should also add some note that the dedicated command requires the server to be stopped and that it requires the same args as start (db options etc.).

I believe this is not resolved yet.

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Admin bootstrap and recovery"
summary="Learn how to bootstrap and recover admin account.">

== A temporary admin account

> A user or service admin account created using one of the methods described below is *temporary*. This means the account should exist only for the duration necessary to perform operations needed to gain permanent and more secure admin access. Various UI/UX elements, such as the Administration Console warning banner, labels, and log messages, will indicate to a {project_name} administrator that the account is temporary.

Reviewer 2: Could you elaborate more on what temporary means here? Is the access available until the Keycloak node that was started with the parameters is alive?

Reviewer: By "temporary" we mean "considered to be temporary by the user". Basically it's just semantical for now and only highlighted as a [warning](https://github.com/keycloak/keycloak/pull/31387) in the UI and logs. In the future, the accounts will be [expired](https://github.com/keycloak/keycloak/issues/30008#issuecomment-2151956072).

Reviewer 2: I see. Should we mention in the docs they need to delete the temporary users manually?

Author: Addressed.

Reviewer 2: Thank you @Pepo48

<@kc.start parameters="--bootstrap-admin-username tmpadm --bootstrap-admin-password pass"/>

<@kc.startdev parameters="--bootstrap-admin-client-id tmpadm --bootstrap-admin-client-secret secret"/>

The username or client ID values can be omitted; see the <<Default values>> section below for more information.

The purpose of these options is solely for bootstrapping temporary admin accounts. These accounts will be created only during the initial start of the {project_name} server when the master realm doesn't exist yet. The accounts are always created in the master realm. For recovering lost admin access, use the dedicated command described in the sections below.

== Bootstrapping an admin user or service account using the dedicated command

> The `bootstrap-admin` command can be executed even before the first-ever start of {project_name}. Bear in mind that the server needs to be stopped prior to using this command. Its execution will trigger the creation of the initial master realm, and as a result, the startup options to bootstrap the admin user and service account will be ignored later when the server is started for the first time.

Reviewer 2: How does this work with Keycloak cluster? Do all nodes need to be stopped before this procedure is executed?

Reviewer: Yes, all nodes need to be stopped:
https://github.com/keycloak/keycloak/blob/1e14b310669126d2dba3d20c902054f3f253ca61/docs/guides/server/bootstrap-admin-recovery.adoc?plain=1#L27

Maybe we could more emphasis on it, make it a warn box or something.

Reviewer 2: For example, changing it to "all {project_name} nodes"?

Author: Addressed as per suggestion.

Reviewer 2: Thank you @Pepo48

Additionally, it is strongly recommended to use the dedicated command with the same options that the {project_name} server is started with (e.g., `db` options).

=== Create an admin user

To create a temporary admin user, execute the following command:

<@kc.bootstrapadmin parameters="user"/>

> If no other parameters are specified and/or no corresponding environment variables are set, the user is prompted to enter the required information. The username or client ID values can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.

Reviewer: ```suggestion
If no other parameters are specified and/or no corresponding environment variables are set, the user is prompted to enter the required information. The username value can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```

Alternatively, the parameters can be directly specified in the command:

<@kc.bootstrapadmin parameters="user --username tmpadm --password:env PASS_VAR"/>

This command creates a temporary admin user with the username `tmpadm` and the password retrieved from the environment variable.

=== Create a service account

In automated scenarios, a temporary admin service account can be a more suitable alternative to a temporary admin user.

To create a temporary admin service account, execute the following command:

<@kc.bootstrapadmin parameters="service"/>

> Similarly, if no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. The username or client ID values can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.

Reviewer: ```suggestion
Similarly, if no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. The client ID value can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```

Alternatively, the parameters can be directly specified in the command:

<@kc.bootstrapadmin parameters="service --client-id tmpclient --client-secret:env=SECRET_VAR"/>

This command creates a temporary admin service account with the client ID `tmpclient` and the secret retrieved from the environment variable.

> == Admin bootstrapping using the {project_name} operator

Reviewer: I'd do it the other way around to keep it consistent with other guides – instead of add links to server guide pointing to the operator guide, I'd suggest adding a link to the operator guide pointing to server guide.

Author: Addressed.

<@tmpl.guide
title="Admin bootstrap and recovery"
summary="Learn how to bootstrap and recover admin account.">

== Creating a temporary admin account

A user account created by one of the methods described below is *temporary*. This means that such an account should exist only for the period necessary to perform operations needed to gain permanent and more secure admin access. Various UI/UX elements such as the Administration Console warning banner, labels or log messages will indicate to a {project_name} administrator that the account is temporary.

== Bootstrapping an admin user using a dedicated command

To create a temporary admin user, execute the following command:

<@kc.bootstrapadmin parameters="user"/>

If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter required information. For more information, see the <<Environment variables>> section bellow.

Alternatively, the parameters can be directly specified in the command:

<@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>

This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.

> NOTE: The command can be executed even before the {project_name} first ever start. However, it leaves the {project_name} server in a state, where the temporary admin account will be the only account that exists in the master realm - the default admin will not be created. If the intention is to recover a lost admin access, the default admin should exist within the master realm prior to the command execution. Also, it is strongly recommended to use the dedicated command with the very sane options that the {project_name} server is started with (e.g. `db` options).

Author: > It'd be also good to add a section for bootstrapping at server startup. Explaining the difference from the dedicated command (that on startup master realm must on exist, making the dedicated command the only suitable option for recovery).

I tried to address this one via a note.
@shawkins wdyt?

> NOTE: The command can be executed even before the first-ever start of {project_name}. However, this will leave the {project_name} server in a state where the temporary admin account is the only account that exists in the master realm and the default admin will not be created. The initial admin creation happens only when the master realm is created. If the intention is to recover lost admin access, the default admin must exist in the master realm prior to executing the command (see <@links.server id="configuration" anchor="create-initial-admin"/>). Additionally, it is strongly recommended to use the dedicated command with the same options that the {project_name} server is started with (e.g., `db` options).

Reviewer 3: IMHO the `db` options would not only be "strongly recommended", but "required" as it would otherwise not write to the correct database and it will not work. 

At the same time I am not sure if all of the options I use on the start command will also work on the boostrap commands (like: setting a HTTP port).

Reviewer 4: This is a somewhat confusing area in general for us #29200 - we do some configuration exclusions for the CLI, but those same options will still work via the other config sources (the operator effectively uses this for export because almost everything is set via an env variable). Only a handful of options have hard-coded settings when running a non-server command.

Reviewer 4: I tried to clarify the cases with https://github.com/keycloak/keycloak/issues/29200#issuecomment-2233350379 - as there are potentially 3 different ways users could be invoking non-server commands which have some nuance that we trying to address here quickly with this recommendation.

== Bootstrapping a temporary admin account at {project_name} startup

{project_name} `start` and `start-dev` commands support options for bootstrapping both temporary admin users and admin service accounts. These options are standard configuration options, so they can be specified in any of the https://www.keycloak.org/server/configuration#_configuring_sources_for_keycloak[configuration sources]. For instance, the following examples demonstrate how to use the `start` and `start-dev` commands with CLI parameters to bootstrap a temporary admin user and an admin service account, respectively:

<@kc.start parameters="--bootstrap-admin-username tmpadm --bootstrap-admin-password pass"/>

<@kc.startdev parameters="--bootstrap-admin-client-id tmpadm --bootstrap-admin-client-secret secret"/>

> The purpose of these options is solely for bootstrapping temporary admin accounts. The accounts are always created in the master realm. For recovering lost admin access, use the dedicated command described in the sections below.

Author: Addressed.

== Bootstrapping an admin user using a dedicated command

To create a temporary admin user, execute the following command:

<@kc.bootstrapadmin parameters="user"/>

If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter the required information. For more information, see the <<Environment variables>> section below.

Alternatively, the parameters can be directly specified in the command:

> <@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>

Reviewer: Note that this applies to service account's secret as well.

> <@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>

Author: You're correct, I overlooked it for the part in question (as for the service the behaviour should be already documented correctly). I will fix this.

> <@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>

Author: Fixed.

This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.

NOTE: The command can be executed even before the first-ever start of {project_name}. However, this will leave the {project_name} server in a state where the temporary admin account is the only account that exists in the master realm and the default admin will not be created. The initial admin creation happens only when the master realm is created. If the intention is to recover lost admin access, the default admin must exist in the master realm prior to executing the command (see <@links.server id="configuration" anchor="create-initial-admin"/>). Additionally, it is strongly recommended to use the dedicated command with the same options that the {project_name} server is started with (e.g., `db` options).

== Bootstrapping an admin service account

In automated scenarios, an admin service account can be a more suitable alternative to a temporary admin user.

To create a temporary admin service account, execute the following command:

<@kc.bootstrapadmin parameters="service"/>

> Similarly, the command can be executed regardless of the `master` realm’s existence. Given the client nature of the admin service account, the default admin can be created even after the command execution. If no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. For more information, see the <<Environment variables>> section below.

Author: What I meant by the statement is that you can create the service first and once you open the admin console, you will be still prompted to create the default admin user account. The formulation is weird mainly because at the time of the writing I considered the default admin and the temporary admin to be some sort of different (thus even the NOTE section you rewrote earlier). Only the UI work made me realize that the default admin is also considered to be temporary. Thanks for noticing this, this needs to be rewritten as well.

> Similarly, the command can be executed regardless of the `master` realm’s existence. Given the client nature of the admin service account, the default admin can be created even after the command execution. If no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. For more information, see the <<Environment variables>> section below.

Author: Addressed by slightly changing the section structure for the dedicated command section of the guide.

== Regaining access to the realm with an increased security

Passwordless, OTP, or other advanced authentication methods can be enforced for a realm with lost admin access. In such a case, the admin service account needs to be created to recover lost admin access to the realm. After the service account is created, authentication against the {project_name} instance is required to perform all necessary operations:

<@kc.admin parameters="config credentials --server http://localhost:8080 --realm master --client <service_account_client_name> --secret <service_account_secret>"/>

Next, retrieve the `credentialId`. For this example, the OTP credential is the relevant one. Use the following command to get an array of `CredentialRepresentation` objects and find the one with `type` set to `otp`:

<@kc.admin parameters="get users/{userId}/credentials -r {realm}"/>

Finally, the retrieved ID can be used to remove the advanced authentication method (in our case, OTP):

<@kc.admin parameters="delete users/{userId}/credentials/{credentialId} -r {realm}"/>

> == Disable the parameters prompt

Author: Addressed.

To disable the prompt for the parameters, the `--no-prompt` parameter can be used. For example:

<@kc.bootstrapadmin parameters="user --username tmpadm --no-prompt"/>

If no corresponding environment variable is set, the command will fail with an error message indicating that the required password parameter is missing.

== Environment variables

> For the `bootstrap-admin user` command, both username and password can be optionally set as environment variables:

Reviewer: > the client ID is optional and defaults to temp-admin-service

Username also has a default value. Please see the comment above.

Author: I believe this was already partially addressed (I just additionally mentioned the requirement for the server to be stopped) - the question is about the appropriate wording here (see Alex's question and Steven's answer above).

As for the default values, I added a new section.

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Admin bootstrap and recovery"

> summary="Learn how to bootstrap and recover admin account." >

Author: As I mentioned during the meeting, initially I tried: 
```suggestion
summary="Learn how to bootstrap and recover admin account." 
includedOptions="bootstrap-admin-*">
```
but the Relevant options table ended up to be empty. Any idea why?

> == A solution for a disaster scenario

Reviewer: Bootstrapping is not just for disaster recovery scenarios.

Author: Adressed.

> It used to be difficult to regain access to a {project_name} instance, where all admin users have been locked out. The process required multiple user-advanced steps, including direct database access and manual changes. In the continuous effort to improve the user experience, {project_name} now provides multiple ways to bootstrap a new admin account, which can be later used to recover from such a situation.

Reviewer: This paragraph is good but possibly should be moved to release notes. IMHO guides should not reminiscence on some previous state of things.

Instead, I'd suggest some introduction that would explain the user the concept of temporary admin user (that they should create a more permanent and secure user etc.).

== Bootstrapping an admin user using a dedicated command

To create a temporary admin user, execute the following command:

<@kc.bootstrapadmin parameters="user"/>

This command can be executed even before the {project_name} first ever start. In other words, it does not matter, whether the `master` realm already exists or not. If no other parameters are specified and/or no corresponding environment variables are set, a user is prompted to enter required information. For more information, see the <<Environment variables>> section bellow.

Alternatively, the parameters can be directly specified in the command:

<@kc.bootstrapadmin parameters="user --username tmpadm --password tmpadmpass"/>

> This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`. The user is valid for 60 minutes.

Reviewer: We do not automatically expire the user at this phase, see https://github.com/keycloak/keycloak/issues/30008#issuecomment-2151956072.
```suggestion
This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.
```

Author: @vmuzikar you probably reviewed an older revision - at the time you posted the review the text didn't contain any reference regarding the expiration. I immediately cleaned it up after our discussion at the meeting.

Thanks for pointing it out in the fist place.

Author: This was a leftover I forgot about. Adressed, thanks!

== Bootstrapping an admin service account

In automated scenarios, an admin service account can be more suitable alternative to a temporary admin user.

Again, to create a temporary admin service account, execute the following command:

<@kc.bootstrapadmin parameters="service"/>

The same rule applies here - the command can be executed regardless of the `master` realm existence. If no corresponding environment variables nor the additional parameter are set, a user will be prompted to enter the required information. For more information, see the <<Environment variables>> section bellow.

Alternatively, you can specify the parameters directly in the command:

<@kc.bootstrapadmin parameters="service --client-id tmpclient --client-secret:env=SECRET_VAR"/>

This command creates a temporary admin service account with the client ID `tmpclient` and the secret retrieved from the environment variable.

== Regaining access to the realm with an increased security

> Passwordless, OTP or other advanced authentication methods can be enforced for the realm with lost admin access. In such case, the admin service account needs to be created. After the service account is created, authentication against the {project_name} instance is required to perform all the necessary operations:

Reviewer: ```suggestion
Passwordless, OTP or other advanced authentication methods can be enforced for the realm with lost admin access. In such case, the admin service account needs to be created to recover lost admin access to such realm. After the service account is created, authentication against the {project_name} instance is required to perform all the necessary operations:
```

Author: Adressed.

[source]
----
POST realms/{realm}/protocol/openid-connect/token \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "client_id=<service_account_id>" \
  -d "client_secret=<service_account_secret>" \

>   -d "grant_type=client_credentials"

Reviewer: We could probably use kcadm.sh|bat in these examples right away.

Author: Adressed.

[source]
----
GET admin/realms/{realm}/users/{userId}/credentials
----

Finally, the retrieved ID can be used to remove the advanced authentication method (in our case OTP):

[source]
----
DELETE admin/realms/{realm}/users/{userId}/credentials/{credentialId}
----

To perform the above operations, `kcadm.sh` utility can be used as well as any HTTP client, like `curl`.

== Disable the parameters prompt

If you want to disable the prompt for the parameters, you can use the `--no-prompt` parameter. For example:

<@kc.bootstrapadmin parameters="user --username tmpadm --no-prompt"/>

If no corresponding environment variable is set, the command will fail with an error message, because the required password parameter is missing.

== Environment variables

For the `bootstrap-admin user` command, both username and password can be optionally set as environment variables:

<@kc.bootstrapadmin parameters="user --username:env=<YourUsernameEnv> --password:env=<YourPassEnv>"/>

> Regarding the `bootstrap-admin service` command, for the client ID it is optional, while for the client secret it is required to be set as an environment variable:

Author: Addressed. Thanks @shawkins.
